synapsebridgehelpers/export_tables.py

```diff
@@ -0,0 +1,342 @@
+            modified_target_columns = list(filter(lambda c : c["name"] in cols,
+                                                  target_cols))
+            for c in modified_target_columns:
+                target_schema.removeColumn(c)
+            target_schema.addColumns(modified_source_columns)
+        if action == "renamed":
+            for target_name, source_name in cols.items():
+                renamed_source_column = next(
+                        filter(lambda c : c["name"] == source_name, source_cols))
+                renamed_target_column = next(
+                        filter(lambda c : c["name"] == target_name, target_cols))
+                target_schema.removeColumn(renamed_target_column)
+                target_schema.addColumn(renamed_source_column)
+    target_schema = syn.store(target_schema)
+    return target_schema
+
+
+def export_tables(syn, table_mapping, target_project=None, update=True,
+                  reference_col="recordId", copy_file_handles=True, **kwargs):
+    """Export records from one Synapse table to another. Or transfer tables
+    to completely new tables in a separate project.
+
+    Parameters
+    ----------
+    syn : synapseclient.Synapse
+    table_mapping : dict, list, or str
+        If exporting records of one or more tables to other, preexisting tables,
+        table_mapping is a dictionary containing Synapse ID key/value mappings
+        from source to target tables. If exporting table records to not yet
+        created tables in a seperate project, table_mapping can be a list or
+        string.
+    target_project : str
+        If exporting table records to not yet created tables in a seperate
+        project, specify the target projects Synapse ID here.
+    update : bool, defaults to True
+        When exporting records of one or more tables to other, preexisting
+        tables, whether to append new records to these tables or completely
+        overwrite the table records.
+    referenceCol : str
+        If `update` is True, use this column as the table index to determine
+        which records are already present in the target table.
+    copy_file_handles : bool, defaults to True
+        Whether to copy the file handles from the source table to the target
+        table. If you are not the creator of these file handles, this must
+        be set to True if you want to a column containing file handles in your
+        target table.
+    **kwargs
+        Additional arguments to pass to synapsebridgehelpers.query_across_tables
+
+    Returns
+    -------
+    """
+    results = {}
+    if isinstance(table_mapping, (list, str)): # export to brand new tables
+        if target_project is None:
+            raise TypeError("If passing a list to table_mapping, "
+                            "target_project must be set.")
+        source_tables = synapsebridgehelpers.query_across_tables(
+                syn, tables=table_mapping, as_data_frame=True, **kwargs)
```

Comment:
only named arguments then.

Reply:
This hides a lot of complexity and makes it harder for someone to review what happens.  It would be more explicit to have the tables passed into this function instead of hiding the query logic at the end of the parameters.  I.e. something like:

```
data_to_copy = query_across_tables(...)
export_tables(..., data_to_copy,...)
```

Reply:
This should be a subroutine of `export_tables`. Let's call it `store_tables`. If we have a function like `store_tables` that takes a list (`data_to_copy`) of pandas DataFrames from who-knows-where, then it doesn't make much sense to have all the additional functionality of synchronizing schemas and replacing file handles. All that should be done before you pass `data_to_copy` to `store_tables`.

The use cases we have so far only require simple queries where some identifier is in a list of other identifiers or we can do a match by substudy. I think I can clean up export_tables.py so that if someone does have a complex query that they need very carefully curated, they can replicate the behavior of `export_tables` with
```
query_across_tables(...)
... # optionally other DataFrame curations
replace_file_handles(...)
synchronize_schemas(...)
store_tables(...)
```

Reply:
If `query_across_tables` returned a dictionary mapping table Synapse IDs to results we could include an optional parameter `source_tables` that would act as a stand in value for the call to `query_across_tables`. But since `query_across_tables` just returns a list with no relation to the input that produced it we can't be sure what we're supposed to update in the case that `table_mapping` is a dictionary.

Reply:
note to self: TODO

look into whether we can include a table_mapping and an argument of the same type as that returned by query_across tables.

Reply:
@larssono What we can do is -- instead of having users pass a list of DataFrames like that returned by `query_across_tables` -- they will pass a dictionary mapping synapse IDs to the new records retrieved from those sources. This in conjunction with the `table_mapping` parameter allows us to correctly map new records to their target tables.

An example call would look like this:
```
tables = list(table_mapping)
new_records = query_across_tables(syn, tables, ...)
source_tables = {t: df for t, df in zip(tables, new_records)}
export_tables(syn, table_mapping, source_tables, ...)
```

Like I whined about in my comment above, if I didn't have to be backwards compatible with the current implementation of `query_across_tables`, `query_across_tables` would create this `source_tables` object for us 😒